### PR TITLE
add solution

### PR DESCRIPTION
**Pull request by dimarogkov**
[DEMO LINK](https://dimarogkov.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/components/TodoList/TodoList.tsx [hunk 1]
```diff
@@ -1 +1,21 @@
...
+import { TodoInfo } from '../TodoInfo';
+
+interface Props {
+  todos: {
+    todoUser: User | undefined;
+    id: number;
+    title: string;
+    completed: boolean;
+    userId: number;
+  }[];
```
Comment (etojeDenys) [inline src/components/TodoList/TodoList.tsx:12]:
```suggestion
  todos: (Todo & {todoUser?: User})[];
```
we can simplify it

File: src/components/Form/Form.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      setUserId(0);
+      setUserIdInvalid(false);
+    }
+  };
+
+  return (
+    <form
+      action="/api/todos"
+      method="POST"
+      onSubmit={(event) => submitForm(event)}
```
Comment (etojeDenys) [inline src/components/Form/Form.tsx:65]:
```suggestion
      onSubmit={submitForm}
```
we can simplify it

**Review by etojeDenys — CHANGES_REQUESTED** [review]
i can create the todo only with spaces, let's fix this case

File: src/interfaces/todo.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface Todo {
+  id: number,
+  title: string,
+  completed: boolean,
+  userId: number,
```
Comment (VitaliyBondarenko1982) [inline src/interfaces/todo.ts:5]:
```suggestion
  userId: number,
  todoUser?: User,
```
and use this type everwhere

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,33 @@
...
-            delectus aut autem
-          </h2>
+    return todos.map(todo => ({ ...todo, todoUser: getUserById(todo.userId) }));
+  };
 
-          <a className="UserInfo" href="mailto:Sincere@april.biz">
-            Leanne Graham
-          </a>
-        </article>
+  const finalTodos = createTodosArr(todosArr, usersFromServer);
```
Comment (VitaliyBondarenko1982) [inline src/App.tsx:23]: better to make this function outside of component for initial todos state and after that change just state of todos.
in your implementation you make this operation during add each new todo

```
const getUserById = (userId: number) => {
  return usersFromServer.find(user => user && user.id === userId) || null;
};

const todosWithUser = todosFromServer
  .map(todo => ({ ...todo, todoUser: getUserById(todo.userId) }));
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,33 @@
...
-        <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
-        </div>
-
-        <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
-          </select>
+  const [todosArr, setTodosArr] = useState<Todo[]>(todosFromServer);
```
Comment (VitaliyBondarenko1982) [inline src/App.tsx:11]:
```suggestion
  const [todosArr, setTodosArr] = useState<Todo[]>(todosWithUser);
```

File: src/components/Form/Form.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    if (title.length > 0 && userId > 0) {
+      setId(prev => prev + 1);
+
+      const newTodo: Todo = {
+        id,
+        title,
+        completed: false,
+        userId,
```
Comment (VitaliyBondarenko1982) [inline src/components/Form/Form.tsx:50]:
```suggestion
      const newTodo: Todo = {
        id,
        title,
        completed: false,
        userId,
        todoUser: getUserById(userId)
```
and here you can add user immediately during make new todo

**Review by VitaliyBondarenko1982 — CHANGES_REQUESTED** [review]
Good job.
Needs some improvements.

**Review by etojeDenys — CHANGES_REQUESTED** [review]
still can add the todo only with spaces
image

**Review by etojeDenys — APPROVED** [review]
good job

**Pull request closed without merging**